### DISCUSSION TITLE
IPSec Transport mode with IPIP Encapsulation?

### DISCUSSION THREAD

**Discussion by opg7371** [strongswan/strongswan#1117] [Q&A]
We have a real server behind the proxy server ( or a firewall).  The client must connect to the real server through IPSec Transport mode Encryption. The IPSec Transport mode works between the two hosts and fails whenever a hop is introduced between the two hosts. This is due to Checksum Integrity failure at the destination which is due to DNAT at the intermittent hop between 2 hosts.

We want something to encapsulate the ESP Transport mode packet either in a IP Header or UDP encapsulation or something like this ->

A -----------------------------------------> GATEWAY -----------------------------> B
Transport mode            (Transport mode is received                    Packet accepted
                                      and IP header in encapsulated
                                      on the packet and is forwarded 
                                      again. )  

Is something like this possible where the Client sends the Transport mode ESP packet and the intermittent hosts encapsulate that packet (either in IP/UDP) and then send it to the destination?
Is there any RFC for these types of scenarios?

Thanks in advance.

**Comment by Thermi** [comment]
> The IPSec Transport mode works between the two hosts and fails whenever a hop is introduced between the two hosts. This is due to Checksum Integrity failure at the destination which is due to DNAT at the intermittent hop between 2 hosts.

Are you using AH instead of ESP? ESP only authenticates the payload on layer 4 and higher, not layer 3. So an ICV failure would only be possible if the intermediate router violates the RFC and mutates the payload without understanding it first.
If they understood it first they wouldn't touch it because they'd see that they can't NAT it. That is because ESP has no port fields to mutate. Only the IP addresses can be mutated (and any variable fields like TTL, DSCP/TOS, ...).

I suppose there's something weird going on.

If NAT is detected strongSwan automatically switches to UDP encapsulation. Then the NAT gateway has ports to fiddle around with. They are not part of the authenticated payload. The UDP header is discarded as part of IPsec processing.

Transport mode is compatible with UDP encapsulation.

**Reply by opg7371** [reply]
Thanks for fast response.

Transport mode is definitely compatible with UDP encapsulation. (NAT-T with port 4500).
But as far as I know, it looks good if the UDP encapsulation was done at the client-side.

What I want to know is that is it possible for some middle gateway to UDP to encapsulate the existing IPSEC ESP Transport mode coming from the client and then forward the request to the destination and vice-versa for the response of the requests.

In our legacy network, we are not using IKE (Dynamic mode), rather we are using the static mode. So there is no way for the client to find out whether he is behind NAT or not.

**Reply by Thermi** [reply]
> Transport mode is definitely compatible with UDP encapsulation. (NAT-T with port 4500).

Yes. If both side use UDP encapsulation. If for some reason UDP encapsulation is not supported it might cause issues.

> What I want to know is that is it possible for some middle gateway to UDP to encapsulate the existing IPSEC ESP Transport mode coming from the client and then forward the request to the destination and vice-versa for the response of the requests.

No, that's not standardized. You can of course program that as a userspace daemon on a router if you desire that.

> In our legacy network, we are not using IKE (Dynamic mode), rather we are using the static mode. So there is no way for the client to find out whether he is behind NAT or not.

I understand.
